script/vsts/platforms/templates/cache.yml

```diff
@@ -11,20 +11,29 @@ steps:
   - task: Cache@2
     displayName: Cache node_modules
     inputs:
-      key: 'npm | "$(Agent.OS)" | "$(BUILD_ARCH)" | package.json, package-lock.json, script/vsts/platforms/${{ parameters.OS }}.yml'
+      key: 'npm | "$(Agent.OS)" | "$(BUILD_ARCH)" | package.json, package-lock.json, script/vsts/platforms/templates/preparation.yml'
+      restoreKeys: |
+         npm | "$(Agent.OS)" | "$(BUILD_ARCH)" | package.json, package-lock.json
```

Comment:
Shouldn't we add the `script/vsts/platforms/templates/preparation.ym` key to restoreKeys as well?

Reply:
`restoreKeys` is typically a subset of `key`.

`restoreKeys` must be a different set of files than `key`, or it will match in exactly the same scenarios and become redundant.

Reply:
I think the goals I have in mind for cache availability are already met on `master`, by the way.

We can speed up in the case of inexact matches by doing:

- Restore `apm/node_modules`
  - Exact cache hit? If yes, then restore `script/node_modules`, if no, skip restoring
    - Exact cache hit for both of the above? If yes, then restore `[repo_root]/node_modules`, if no then skip restoring.

This is to prevent "restore-delete-reinstalls" that waste time.

Reply:
> `restoreKeys` is typically a subset of `key`.

> `restoreKeys` must be a different set of files than `key`, or it will match in exactly the same scenarios and become redundant.

> (The idea is for `restoreKeys` to be looser, so caching can fall back to the looser requirements of `restoreKeys`, if matching to `key` fails.)

I see. thanks for the explanation.


So it brings another question. Shouldn't we add bootstrap to the Keys then? That affects the cache obviously.

Reply:
> Restore apm/node_modules
> Exact cache hit? If yes, then restore script/node_modules, if no, skip restoring

The keys are independent here. apm/package.json and script/package.json are different files and so their cache. We should not decide about independent caches based on other steps.

Reply:
> Shouldn't we add bootstrap to the Keys then?

Hmm I think that's right, in a way. But there's a lot of other stuff maybe we should add then. Upstream isn't doing it and they've been okay so far... Probably because `bootstrap` doesn't change that often. And well... maybe everything in `script/lib` should be in it? Everything in `script/`? Where do you draw the line? I'm not sure.

> The keys are independent here. apm/package.json and script/package.json [...]

Recall that we do not skip bootstrapping if there is _any_ inexact match. And `script/bootstrap` _already_ uses `npm ci` in the CI environment. That deletes the `node_modules` that was restored for the directory it is installing in. (That can lead to restoring a cache and then immediately deleting it. Either of those behaviors can be changed of course). But if we keep those two behaviors, restoring additional caches when one cache hit is inexact is a waste of time.

There are some things we have to take into account IMO:
- Handle Node updates in CI properly (I think having `templates/preparation` in the `key` and `restoreKeys` is enough)
- Not waste time restoring then immediately deleting a cache (un-optimal use of caching + `npm ci`)
- Figure out which is optimal for speed, `npm install` or `npm ci`, given the caches we want to restore. (Without introducing bugs by being too optimistic.)
- Figure out how loose we can make the `restoreKeys` without running into problems.

Reply:
Sorry, I shouldn't have deleted `script/vsts/platforms/${{ parameters.OS }}.yml`. I was being too optimistic and not thinking it through. If we update the core OS, then the compilers might change, or dynamic libraries (?), especially on macOS and Linux, and I guess we shouldn't mix compiler major versions.

Reply:
I should have known that `npm ci` deletes things. If that's the case, this pull request is useless. No matter what is restored bootstrap will delete all of it.

https://github.com/atom-ide-community/atom/blob/4f01cb186d007e2d948de2f9b7e057b3b7137c88/script/lib/run-apm-install.js#L14